File: Task/Simple-database/Java/simple-database.java
import java.io.*;
import java.text.*;
import java.util.*;

public class SimpleDatabase {

    final static String filename = "simdb.csv";

    public static void main(String[] args) {
        if (args.length < 1 || args.length > 3) {
            printUsage();
            return;
        }

        switch (args[0].toLowerCase()) {
            case "add":
                addItem(args);
                break;
            case "latest":
                printLatest(args);
                break;
            case "all":
                printAll();
                break;
            default:
                printUsage();
                break;
        }
    }

    private static class Item implements Comparable<Item>{
        final String name;
        final String date;
        final String category;

        Item(String n, String d, String c) {
            name = n;
            date = d;
            category = c;
        }

        @Override
        public int compareTo(Item item){
            return date.compareTo(item.date);
        }

        @Override
        public String toString() {
            return String.format("%s,%s,%s%n", name, date, category);
        }
    }

    private static void addItem(String[] input) {
        if (input.length < 2) {
            printUsage();
            return;
        }
        List<Item> db = load();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        String date = sdf.format(new Date());
        String cat = (input.length == 3) ? input[2] : "none";
        db.add(new Item(input[1], date, cat));
        store(db);
    }

    private static void printLatest(String[] a) {
        List<Item> db = load();
        if (db.isEmpty()) {
            System.out.println("No entries in database.");
            return;
        }
        Collections.sort(db);
        if (a.length == 2) {
            for (Item item : db)
                if (item.category.equals(a[1]))
                    System.out.println(item);
        } else {
            System.out.println(db.get(0));
        }
    }

    private static void printAll() {
        List<Item> db = load();
        if (db.isEmpty()) {
            System.out.println("No entries in database.");
            return;
        }
        Collections.sort(db);
        for (Item item : db)
            System.out.println(item);
    }

    private static List<Item> load() {
        List<Item> db = new ArrayList<>();
        try (Scanner sc = new Scanner(new File(filename))) {
            while (sc.hasNext()) {
                String[] item = sc.nextLine().split(",");
                db.add(new Item(item[0], item[1], item[2]));
            }
        } catch (IOException e) {
            System.out.println(e);
        }
        return db;
    }

    private static void store(List<Item> db) {
        try (FileWriter fw = new FileWriter(filename)) {
            for (Item item : db)
                fw.write(item.toString());
        } catch (IOException e) {
            System.out.println(e);
        }
    }

    private static void printUsage() {
         System.out.println("Usage:");
         System.out.println("  simdb cmd [categoryName]");
         System.out.println("  add     add item, followed by optional category");
         System.out.println("  latest  print last added item(s), followed by "
                 + "optional category");
         System.out.println("  all     print all");
         System.out.println("  For instance: add \"some item name\" "
                 + "\"some category name\"");
    }
}

File: Task/Simple-database/Nim/simple-database.nim
import algorithm, json, os, strformat, strutils, times

const FileName = "simdb.json"

type

  Item = object
    name: string
    date: string
    category: string

  Database = seq[Item]

  DbError = object of CatchableError


proc load(): Database =
  if fileExists(FileName):
    let node = try: FileName.parseFile()
               except JsonParsingError:
                 raise newException(DbError, getCurrentExceptionMsg())
    result = node.to(DataBase)


proc store(db: Database) =
  try:
    FileName.writeFile $(%* db)
  except IOError:
    quit "Unable to save database.", QuitFailure


proc addItem(args: seq[string]) =
  var db = try: load()
           except DbError: quit getCurrentExceptionMsg(), QuitFailure

  let date = now().format("yyyy-MM-dd HH:mm:ss")
  let cat = if args.len == 2: args[1] else: "none"
  db.add Item(name: args[0], date: date, category: cat)
  db.store()


proc printLatest(args: seq[string]) =
  let db = try: load()
           except DbError: quit getCurrentExceptionMsg(), QuitFailure
  if db.len == 0:
    echo "No entries in database."
    return

  # No need to sort db as items are added chronologically.
  if args.len == 1:
    var found = false
    for item in reversed(db):
      if item.category == args[0]:
        echo item
        found = true
        break
    if not found:
      echo &"There are no items for category '{args[0]}'"
  else:
    echo db[^1]


proc printAll() =
  let db = try: load()
           except DbError: quit getCurrentExceptionMsg(), QuitFailure
  if db.len == 0:
    echo "No entries in database."
    return
  for item in db:
    echo item


proc printUsage() =
  echo &"""
Usage:
  {getAppFilename().splitPath().tail} cmd [categoryName]

  add     add item, followed by optional category
  latest  print last added item(s), followed by optional category
  all     print all

  For instance: add "some item name" "some category name"
  """
  quit QuitFailure


if paramCount() notin 1..3: printUsage()

var params = commandLineParams()
let command = params[0].toLowerAscii
params.delete(0)
case command
of "add":
  if params.len == 0: printUsage()
  addItem(params)
of "latest":
  if params.len > 1: printUsage()
  printLatest(params)
of "all":
  if params.len != 0: printUsage()
  printAll()
